pkg/bgpv1/agent/controller.go

```diff
@@ -221,33 +250,45 @@ func NewController(ctx context.Context, clientset client.Clientset, rtMgr BGPRou
 	}
 
 	// setup policy lister and informer.
-	policyLister := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Lister()
-	policyInformer := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Informer()
-	policyInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+	c.PolicyLister = factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Lister()
+	c.policyInformer = factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Informer()
+	c.policyInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc:    sig.Event,
 		UpdateFunc: func(_ interface{}, _ interface{}) { sig.Event(struct{}{}) },
 		DeleteFunc: sig.Event,
 	})
-	go policyInformer.Run(stop)
-
-	c := Controller{
-		NodeSpec:     nodeSpecer,
-		PolicyLister: policyLister,
-		Sig:          sig,
-		BGPMgr:       rtMgr,
-	}
 
 	// apply any options.
-	for _, opt := range opts {
+	for _, opt := range params.Opts {
 		opt(&c)
 	}
 
-	// stop channel is passed here, if ctx is canceled the passed stop chan
-	// is closed as well.
-	go c.Run(ctx, stop)
+	params.LC.Append(&c)
+
 	return &c, nil
 }
 
+// Start is called by hive after all of our dependencies have been started.
+func (c *Controller) Start(_ hive.HookContext) error {
+	go c.policyInformer.Run(c.ctx.Done())
+	go c.NodeSpec.Run(c.ctx)
+	go c.Run()
+	return nil
+}
+
+// Stop is called by hive upon shutdown, after all of our dependants have been stopped.
+// We should perform a graceful shutdown and return as soon as done or when the stop context is done.
+func (c *Controller) Stop(ctx hive.HookContext) error {
+	c.cancel()
+
+	select {
+	case <-ctx.Done():
+	case <-c.doneChan:
```

Comment:
What about waiting for `policyInformer.Run` and `NodeSpec.Run`? Consider using workerpool in follow-up PR.